### ISSUE TITLE
FixedStorage / StorageResizable

### ISSUE THREAD

**Issue by tower120** [CAD97/storages-api#5]
Hi! Someone on reddit [mentioned your storages-api](https://www.reddit.com/r/rust/comments/viiw79/comment/idek6xv/?utm_source=share&utm_medium=web2x&context=3).

I made similar storage/mem interface for my [any_vec lib](https://docs.rs/any_vec/0.9.0/any_vec/index.html#membuilder). I called it MemBuilder. MemBuilder constructs Mem, Mem then can be used as your Storage.

The thing I wanted to point out, is that I found it useful to split Mem on Mem + MemResizable and MemBuilder on MemBuilder + MemBuilderResizable. That allowed my Vec implementation to opt-out size-change operations whenever provided Mem implementation does not support them. For example, `reserve_exact` have no meaning for Stack allocated memory (InlineStorage of yours), and [thanks to the type knowledge we can opt out it compile time](https://docs.rs/any_vec/0.9.0/src/any_vec/any_vec.rs.html#197-201), without making separate implementation for it. Same thing with [with_capacity](https://docs.rs/any_vec/0.9.0/any_vec/struct.AnyVec.html#method.with_capacity).

In your api you could move out `grow` and `shrink` into some `StorageResizable`. That would allow `storage-api` consumers to opt-out operations compile time in their container implementations. Or at least make marker trait for that.

Just wanted to share my "experience" from similar design....

**Comment by CAD97** [comment]
It's definitely worth noting as an alternative. (I hope to bring this to a T-libs Major Change Proposal by the end of the year, and have a Rust Foundation grant towards doing so 🎉)

However, one of the first bits of feedback I got when I shared this on the development Zulip was to reduce complexity. The traits at that time *were* much more complicated as they were still [a typed API](https://github.com/CAD97/storages-api/pull/3), but I landed on the current four trait design because lacking each is a *fundamental limitation* implied by certain types of storages we want to support.

Saying inline storage cannot support `grow` is actually *incorrect*. If you have a storage that has 128 bytes of space and request a handle with layout with size 64, you'll get a handle to the first 64 bytes. If you later ask to grow the handle from 64 bytes to 128 bytes, that's fine! There's space right there, already reserved.

The notable thing your `Mem` knows that my `Handle` doesn't is the layout it handles. Every time you go to resolve a handle, you have to again provide the exact same layout you used to allocate it. I suppose since I resolve to `&[MaybeUninit<u8>]` the length of that could communicate the spare capacity of a single storage... but creating the slice is the exact same cost to assemble it with the passed size or the storage size, so I currently prefer to always return the exact requested size. (There's also a theoretical purity to the function being `[ptr, size, align] -> [ptr, size]` (i.e. just discard the last argument) over `[ptr, size, align] -> [ptr, N]`.)

This does suggest though that some sort of capacity hint is useful, so e.g. the case of growing an inline `Vec` with capacity 8 but an inline storage capacity of 12 doesn't just try resizing to 16 and fail, but has a chance to instead grow to 12.

If you're on (or get on) the [Rust lang development Zulip](https://rust-lang.zulipchat.com/) and tell me your handle there, I'll ping you when I first propose the initiative there.

**Comment by tower120** [comment]
> Saying inline storage cannot support grow is actually incorrect. If you have a storage that has 128 bytes of space and request a handle with layout with size 64, you'll get a handle to the first 64 bytes. If you later ask to grow the handle from 64 bytes to 128 bytes, that's fine! There's space right there, already reserved.

But isn't that artificial abstraction? You actually do have fixed-sized array. This is if you'd ask array to grow.... It's just fixed-sized memory storage.
I agree that would be OK for allocator, but not for storage. IMO.

> The notable thing your Mem knows that my Handle doesn't is the layout it handles.

Emm... That is because my vector is untyped... And allocator requires `Layout` to work, not just size. And that is Element layout, not array.... So - you can think of it as AnyVec specific...

> However, one of the first bits of feedback I got when I shared this on the development Zulip was to reduce complexity. The traits at that time were much more complicated as they were still https://github.com/CAD97/storages-api/pull/3,

Oh I see then. Yes, your initial design is closer to what I would expect from Mem/Storage.

I initially wanted to say that allocate/deallocate is a little but unsafe. And what happens if InlinedStorage will be asked allocate twice? And wouldn't RAII solve that better? But non-clonable Handle do the same thing, actually.

> If you're on (or get on) the [Rust lang development Zulip](https://rust-lang.zulipchat.com/) and tell me your handle there, I'll ping you when I first propose the initiative there.

Nah, thanks, but I'm not into Rust development. I mean, there's just nothing I could do there...

----

To summarize - I see that `storages-api` initial design was close to mine, but was changed under weight of critique. Case solved! :) Thanks for answers.

**Comment by CAD97** [comment]
> But isn't \[`grow`ing handles in fixed storages] artificial abstraction?

Not really: it's *unification* with allocation which doesn't have any usable slack space. If a heap `realloc` doesn't move to a new allocation, it's doing the exact same thing.

> ... non-clonable Handle ...

Ooh, that's another thing I overlooked that makes your `Mem` different: `Mem` is not copy/clone, whereas a `Handle` is always `Copy`. (`Handle` is an abstraction of a raw pointer.)

It should still be reasonable to get your `Mem` API on top of `Storage`, and now I think I understand what you're providing enough to draft a possible example implementation:

(Obvious draft/example elisions, i.e. error handling, `unsafe` blocks, trivial derives)

```rust
struct MemBuilder<S>;

impl MemBuilder<S: Storage + Default> {
    fn build(&mut self, element_layout: Layout) -> Mem {
        let len = 0; // FIXME: prealloc inline
        let mut storage = S::default();
        let handle = storage.allocate(element_layout.repeat(len)?.0)?;
        Mem { storage, handle, element_layout, len }
    }
}

struct Mem<S: Storage> {
    storage: S,
    handle: S::Handle,
    element_layout: Layout,
    len: usize,
}

impl Mem<S: Storage> {
    fn as_ptr(&self) -> *const u8 {
        self.storage.resolve(self.handle, self.full_layout()).as_ptr() as _
    }

    fn as_mut_ptr(&mut self) -> *mut u8 {
        self.storage.resolve_mut(self.handle, self.full_layout()).as_ptr() as _
    }

    fn element_layout(&self) -> Layout {
        self.element_layout
    }

    fn full_layout(&self) -> Layout {
        self.element_layout.repeat(self.len)?.0
    }

    fn size(&self) -> usize {
        self.len
    }

    fn expand_exact(&mut self, additional: usize) {
        let grow_layout = self.element_layout.repeat(self.len + additional)?.0;
        self.handle = self.storage.grow(self.handle, self.full_layout(), grow_layout);
        self.len += additional;
    }
}

impl Drop for Mem<S: Storage> {
    fn drop(&mut self) {
        self.storage.deallocate(self.handle, self.full_layout());
    }
}
```

I think the only thing missing is the FIXME: a way to eagerly get the maximum out of a pre-allocated storage up front. I'll definitely think about that. (Providing the interface in a refinement `trait FixedStorage` would be inverted from what your API refinement wants, though...)

Initial thoughts: the current `Allocator` API allows for spare capacity allocation because it returns essentially[^1] `{ ptr: *mut u8, usable: Layout }`, and it allows deallocation with any layout that “fits” in `requested..=usable`. Doing the same for `Storage` means returning `(Handle, Layout)` from allocation, and using the same “fit” rules for deallocation.

[^1]: It actually returns `ptr::NonNull<[u8]>`. Usable size is the length of the slice, and usable alignment is the alignment of the pointer.

**Comment by tower120** [comment]
> Not really: it's unification with allocation which doesn't have any usable slack space

From this point of view - it is OK. But does this _unification_ desirable?

> Ooh, that's another thing I overlooked that makes your Mem different: Mem is not copy/clone, whereas a Handle is always Copy. (Handle is an abstraction of a raw pointer.)

Speaking of which - even if Handle would unclonable, does allocate/deallocate is somehow better then RAII?

> (Obvious draft/example elisions, i.e. error handling, unsafe blocks, trivial derives)

what is `Mem::full_layout` for? Or that just implementation details?

> Doing the same for Storage means returning (Handle, Layout) from allocation, and using the same “fit” rules for deallocation. And I think in your current design Handle differently should be unique, to prevent double free "events" by design.

Why it should follow allocator interface at all? Is that due some requirement?

**Comment by tower120** [comment]
Another thing I would like to mention, is that in your current design moving 
`expand` inside Storage, and thus controlling grow strategy, would allow to do additional optimizations. For example `SmallMem`, reaching inline storage capacity,  may want to grow to that inline capacity limit, and avoid allocation for now, instead of making allocation just because its capacity pow2 time.

**Comment by CAD97** [comment]
> But does this unification desirable?

I think you can argue for either direction.

My position is that `InlineStorage` *absolutely can* service a request for a smaller storage than it is. From this perspective, `InlineStorage` has the extra ability of *knowing its maximum allocation size*.

> what is `Mem::full_layout` for? Or that just implementation details?

Just implementation.

> Why it should follow allocator interface at all? Is that due some requirement?

No specific constraint; it's just an example related API which provides excess space.

> Another thing I would like to mention, is that in your current design moving
`expand` inside Storage, and thus controlling grow strategy, would allow to do additional optimizations.

It's worth considering, but I'm more inclined that growth strategies should be the job of a `RawVec<T, S>`, which itself is a fancy `Box<[MaybeUninit<T>], S>`. `Storage` should expose enough that `RawVec` can be however smart it wants to be — this could be the available inline space, or generalizing a bit, the largest guaranteed-success allocation size.

A *more eager* growth strategy could be done internally if `Storage` allows returning extra space, but not a *less eager* one.

> Speaking of which - even if Handle would unclonable, does allocate/deallocate is somehow better then RAII?

What this gives us is [`MultipleStorage`](https://cad97.github.io/storages-api/storage_api/trait.MultipleStorage.html), a storage which can allocate multiple handles at the same time.

In order to have RAII on `Handle`, it has to have a reference to the `Storage` in order to deallocate.

Note that we don't have `MemBuilder = Storage, Mem = Handle`. What we instead is closer to this conversion table:

Mem | Storage
- | -
MemBuilder | FnMut(Layout) -> (Storage, Handle)
Mem | (Storage, Handle) + RAII deallocate
() | Handle

Consider the case of `Storage = BumpAllocator, Handle = usize`, roughly

```rust
let mut arena = BumpAllocator::from(vec![0; 0x10_0000]);

let a = arena.allocate(…);
let b = arena.allocate(…);
let c = arena.allocate(…);
let d = arena.allocate(…);
…
```

Each handle is now just `*mut u8` instead of needing to be (*minimally*, for *this case*) `(&mut BumpAllocator, *mut u8)`.

So the use case is multiple handles sharing a storage. The equivalent to MemBuilder is making multiple Storage.

A `Handle` type of `()` is not uncommon. If a `Storage` is not `MultipleStorage`, it's expected! So a better analogous implementation of `Mem` would set `S: Storage<Handle=()>`; this is more restrictive but only in ways not usable by the `MemBuilder`/`Mem` interface.

**Comment by tower120** [comment]
> What this gives us is [MultipleStorage](https://cad97.github.io/storages-api/storage_api/trait.MultipleStorage.html), a storage which can allocate multiple handles at the same time.
> ...
> Each handle is now just *mut u8 instead of needing to be (minimally, for this case) (&mut BumpAllocator, *mut u8).

Hmmm... Yes, I see - makes sense now to me. I don't have enough expertise to judge whether RAII safety worth addition pointer per Storage overhead. Probably there is some use-cases with massive small-sized Storages. After all this design with Handlers, can be wrapped into some RAII interface.
